src/coord/src/coord.rs

```diff
@@ -301,14 +300,28 @@ where
     /// located after `ts_tx`.
     _timestamper_thread_handle: JoinOnDropHandle<()>,
     metric_scraper: Scraper,
-    /// The last timestamp we assigned to a read.
-    read_lower_bound: Timestamp,
-    /// The timestamp that all local inputs have been advanced up to.
-    closed_up_to: Timestamp,
-    /// Whether or not the most recent operation was a read.
-    last_op_was_read: bool,
-    /// Whether we need to advance local inputs (i.e., did someone observe a timestamp).
-    need_advance: bool,
+
+    /// The last known timestamp that was considered "open" (i.e. where writes
+    /// may occur). However, this timestamp is _not_ open when
+    /// `read_at_offset_from_open == 0`; in this case, reads will occur at
+    /// `last_open_local_ts`, and the Coordinator must open a new timestamp
+    /// for writes.
+    ///
+    /// Indirectly, this value aims to represent the Coordinator's desired value
+    /// for `upper` for table frontiers, as long as we know it is open.
+    last_open_local_ts: Timestamp,
+    /// The "distance" away from `last_open_local_ts` to perform reads; will
+    /// either be 0 (when we have pending writes at the open time) or 1
+    /// (otherwise).
+    ///
+    /// Indirectly, this value aims to express the "immediately readable" values
+    /// adjacent to `upper` (one could think of it as `upper--`), where we
+    /// expect reads to be able to return immediately.
+    read_at_offset_from_open_local: u64,
```

Comment:
I have a preference that this is an enum with perhaps a `To<u64>` impl. That would more explicitly define exactly what this thing does, since this was the most confusing part of the PR. I think I get it now, because we need to record three pieces of information: 1) some timestamp, 2) was the last thing a read or write, 3) do we need to advance. This is the third thing, but because of how it affects read timestamps it's sometimes easier to have it be a u64, but that obscures the problem it is solving.

Reply:
Rather than a two-state enum, I just made this a bool. I thought the u64 was a little more illustrative, but sounds like it hindered the code's legibility. Sorry for making this harder than it needed to be.